api/auth/auth.cpp

```diff
@@ -232,47 +234,187 @@ error::Error FetchJWTToken(
+				AuthData auth_data {*server_url_, *token_};
+				// TODO: std::move()
+				ex_auth_data = ExpectedAuthData(auth_data);
+			}
+			PostPendingActions(ex_auth_data);
+		});
+
+	watching_token_signal_ = (err == error::NoError);
+	return err;
 }
 
-void Authenticator::RunPendingActions(ExpectedToken ex_token) {
+void Authenticator::PostPendingActions(ExpectedAuthData ex_auth_data) {
 	for (auto action : pending_actions_) {
-		loop_.Post([action, ex_token]() { action(ex_token); });
+		loop_.Post([action, ex_auth_data]() { action(ex_auth_data); });
 	}
 	pending_actions_.clear();
 }
 
+error::Error Authenticator::RequestNewToken(optional<AuthenticatedAction> opt_action) {
+	if (token_fetch_in_progress_) {
+		// Just make sure the action (if any) is called once the token is
+		// obtained.
+		if (opt_action) {
+			pending_actions_.push_back(*opt_action);
+		}
+		return error::NoError;
+	}
+
+	auto err = dbus_client_.CallMethod<expected::ExpectedBool>(
+		"io.mender.AuthenticationManager",
+		"/io/mender/AuthenticationManager",
+		"io.mender.Authentication1",
+		"FetchJwtToken",
+		[this](expected::ExpectedBool ex_value) {
+			if (!ex_value) {
+				token_fetch_in_progress_ = false;
+				mlog::Error("Failed to request new token fetching: " + ex_value.error().String());
+				ExpectedAuthData ex_auth_data = expected::unexpected(ex_value.error());
+				PostPendingActions(ex_auth_data);
+			} else if (!ex_value.value()) {
+				// mender-auth encountered an error not sent over DBus (should never happen)
+				token_fetch_in_progress_ = false;
+				mlog::Error(
+					"Failed to request new token fetching (see mender-auth logs for details)");
+				ExpectedAuthData ex_auth_data = expected::unexpected(MakeError(
+					AuthenticationError, "Failed to request new token fetching from mender-auth"));
+				PostPendingActions(ex_auth_data);
+			}
+		});
+	if (err != error::NoError) {
+		// A sync DBus error.
+		mlog::Error("Failed to request new token fetching: " + err.String());
+		token_fetch_in_progress_ = false;
+		ExpectedAuthData ex_auth_data = expected::unexpected(err);
+		PostPendingActions(ex_auth_data);
+		if (opt_action) {
+			(*opt_action)(ex_auth_data);
+		}
+		return err;
+	}
+	// else everything went OK
+
+	token_fetch_in_progress_ = true;
+
+	// Make sure the action (if any) is called once the token is
+	// obtained.
+	if (opt_action) {
+		pending_actions_.push_back(*opt_action);
+	}
+
+	// Make sure we don't wait for the token forever.
+	auth_timeout_timer_.AsyncWait(auth_timeout_, [this](error::Error err) {
+		if (err.code == make_error_condition(errc::operation_canceled)) {
+			return;
+		} else if (err == error::NoError) {
+			mlog::Warning("Timed-out waiting for a new token");
+			token_fetch_in_progress_ = false;
+			ExpectedAuthData ex_auth_data = expected::unexpected(
+				MakeError(AuthenticationError, "Timed-out waiting for a new token"));
+			PostPendingActions(ex_auth_data);
+		} else {
+			// should never happen
+			assert(false);
```

Comment:
Q: Should we set `token_fetch_in_progress = false` here, or are we guaranteed that the `opt_action` always handles it?

Reply:
`opt_action` doesn't handle it. It's not super-clear what to do here -- token fetching will still be in progress if the timer fails like this (which it shouldn't). OTOH, extra calls of `FetchJwtToken()` or `GetJwtToken()` are not that much of a big deal. And it's **safer** to set `token_fetch_in_progress = false` here, that's for sure. :+1: